web/packages/teleport/src/Console/DocumentSsh/useFileTransfer.ts

```diff
@@ -46,74 +52,184 @@ export const useFileTransfer = (
+            moderatedSessionId: request.sid,
+          }),
+      });
+    },
+    [currentUser.username, download, filesStore, upload]
+  );
+
+  // handleFileTransferUpdate is called when a FILE_TRANSFER_REQUEST event is received. This is used when
+  // we receive a new file transfer request, or when a request has been updated with an approval but its policy isn't
+  // completely approved yet. An update in this way generally means that the approver array is updated.
+  function handleFileTransferUpdate(data: FileTransferRequest) {
+    setFileTransferRequests(prevstate => {
+      // We receive the same data type when a file transfer request is created and
+      // when an update event happens. Check if we already have this request in our list. If not
+      // in our list, we add it
+      const foundRequest = prevstate.find(
+        ft => ft.requestID === data.requestID
+      );
+      if (!foundRequest) {
+        return [...prevstate, data];
+      } else {
+        return prevstate.map(ft => {
+          if (ft.requestID === data.requestID) {
+            return data;
+          }
+          return ft;
+        });
+      }
+    });
+  }
+
   useEffect(() => {
     // the tty will be init outside of this hook, so we wait until
     // it exists and then attach file transfer handlers to it
-    if (tty) {
-      tty.on(EventType.FILE_TRANSFER_REQUEST, updateFileTransferRequests);
-      tty.on(
-        EventType.FILE_TRANSFER_REQUEST_APPROVE,
-        handleFileTransferApproval
-      );
-      tty.on(EventType.FILE_TRANSFER_REQUEST_DENY, handleFileTransferDenied);
+    if (!tty) {
+      return;
     }
-
+    tty.on(EventType.FILE_TRANSFER_REQUEST, handleFileTransferUpdate);
+    tty.on(EventType.FILE_TRANSFER_REQUEST_APPROVE, handleFileTransferApproval);
+    tty.on(EventType.FILE_TRANSFER_REQUEST_DENY, handleFileTransferDenied);
     return () => {
-      if (tty) {
-        tty.removeListener(
-          EventType.FILE_TRANSFER_REQUEST,
-          updateFileTransferRequests
-        );
-        tty.removeListener(
-          EventType.FILE_TRANSFER_REQUEST_APPROVE,
-          handleFileTransferApproval
-        );
-        tty.removeListener(
-          EventType.FILE_TRANSFER_REQUEST_DENY,
-          handleFileTransferDenied
-        );
+      if (!tty) {
+        return;
       }
+      tty.removeListener(
+        EventType.FILE_TRANSFER_REQUEST,
+        handleFileTransferUpdate
+      );
+      tty.removeListener(
+        EventType.FILE_TRANSFER_REQUEST_APPROVE,
+        handleFileTransferApproval
+      );
+      tty.removeListener(
+        EventType.FILE_TRANSFER_REQUEST_DENY,
+        handleFileTransferDenied
+      );
     };
-  }, [tty]);
+  }, [tty, handleFileTransferDenied, handleFileTransferApproval]);
+
+  /*
+   * Transfer handlers
+   */
 
   function removeFileTransferRequest(requestId: string) {
     setFileTransferRequests(prevstate =>
       prevstate.filter(ft => ft.requestID !== requestId)
     );
   }
```

Comment:
This should be in the section above, shouldn't it? It's not used in transfer handlers.